File: Design-Patterns-Using-Go/Bridge/BridgePatternShape.go
package main

import "fmt"

// Color interface (Implementor)
type Color interface {
	Fill() string
}

// Shape interface (Abstraction)
type Shape interface {
	Draw()
}

// Rectangle struct
type Rectangle struct {
	color Color
}

// NewRectangle constructor
func NewRectangle(color Color) *Rectangle {
	return &Rectangle{color: color}
}

// Draw method for Rectangle
func (r *Rectangle) Draw() {
	fmt.Printf("Drawing Rectangle with color %s\n", r.color.Fill())
}

// Circle struct
type Circle struct {
	color Color
}

// NewCircle constructor
func NewCircle(color Color) *Circle {
	return &Circle{color: color}
}

// Draw method for Circle
func (c *Circle) Draw() {
	fmt.Printf("Drawing Circle with color %s\n", c.color.Fill())
}

// Red struct
type Red struct{}

// Fill method for Red
func (r *Red) Fill() string {
	return "Red"
}

// Green struct
type Green struct{}

// Fill method for Green
func (g *Green) Fill() string {
	return "Green"
}

// Blue struct
type Blue struct{}

// Fill method for Blue
func (b *Blue) Fill() string {
	return "Blue"
}

// Client Code
func main() {
	c1 := &Red{}
	circle := NewCircle(c1)
	circle.Draw() // Output: Drawing Circle with color Red

	c2 := &Green{}
	rectangle := NewRectangle(c2)
	rectangle.Draw() // Output: Drawing Rectangle with color Green
}

/*
Drawing Circle with color Red
Drawing Rectangle with color Green
*/

File: Design-Patterns-Using-Cpp/Bridge/BridgePatternShape.cpp
#include <iostream>

// Implementor interface
class Color {
public:
    virtual std::string fill() = 0;
    virtual ~Color() {} // Virtual destructor for proper cleanup
};

// Red class
class Red : public Color {
public:
    std::string fill() override {
        return "Red";
    }
};

// Green class
class Green : public Color {
public:
    std::string fill() override {
        return "Green";
    }
};

// Blue class
class Blue : public Color {
public:
    std::string fill() override {
        return "Blue";
    }
};

// Abstraction interface
class Shape {
protected:
    Color* imp;

public:
    Shape(Color* imp) : imp(imp) {}

    virtual void draw() = 0;
    virtual ~Shape() {} // Virtual destructor for proper cleanup
};

// Rectangle class
class Rectangle : public Shape {
public:
    Rectangle(Color* imp) : Shape(imp) {}

    void draw() override {
        std::cout << "Drawing Rectangle with color " << imp->fill() << std::endl;
    }
};

// Circle class
class Circle : public Shape {
public:
    Circle(Color* imp) : Shape(imp) {}

    void draw() override {
        std::cout << "Drawing Circle with color " << imp->fill() << std::endl;
    }
};

// Client code
int main() {
    Color* c1 = new Red();
    Shape* abstraction = new Circle(c1);
    abstraction->draw();

    Color* c2 = new Green();
    Shape* abstraction2 = new Rectangle(c2);
    abstraction2->draw();

    // Clean up memory
    delete c1;
    delete abstraction;
    delete c2;
    delete abstraction2;

    return 0;
}

/*
Drawing Circle with color Red
Drawing Rectangle with color Green
*/
